### PR TITLE
[IMP] mail: Edited label on edited message

### PR DESCRIPTION
**Pull request by maki-odoo**
By this commit, if a user edits the message body or deletes an attachment from the message, the _edited_ label in the blob, shows the last edit Info.

Task-2664848

https://github.com/odoo/enterprise/pull/40208

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/static/src/core/message_model.js [hunk 1]
```diff
@@ -171,8 +174,11 @@ export class Message {
 
     get isBodyEmpty() {
         return (
-            !this.body ||
-            ["", "<p></p>", "<p><br></p>", "<p><br/></p>"].includes(this.body.replace(/\s/g, ""))
+            !(this.editDate && this.attachments.length) &&
+            (!this.body ||
+                ["", "<p></p>", "<p><br></p>", "<p><br/></p>"].includes(
+                    this.body.replace(/\s/g, "")
+                ))
```
Comment (alexkuhn) [inline addons/mail/static/src/core/message_model.js:181]: This diff doesn't look correct: `isBodyEmpty` should only have condition on `this.body`.
Shouldn't this be in `isEmpty` instead?

Comment (maki-odoo) [reply addons/mail/static/src/core/message_model.js:181]: We use `isBodyEmpty` condition on showing the message content. Maybe with this diff, the name of function doesn't look quite good, but I think the usage of `isEmpty` is different. Actually with this extra condition, we still have _`edited`_ in the blob, in case of message body is deleted completely and there is a attachment, which visually seems like we have something in the message body.

Comment (alexkuhn) [reply addons/mail/static/src/core/message_model.js:181]: It's safer to make another function `hasTextContent`, so that you distinguish both semantically.

Comment (alexkuhn) [reply addons/mail/static/src/core/message_model.js:181]: For example, in showing of reply to, we don't want to see `(edited)`: we want to inline preview the actual body content of parent message.

File: addons/mail/models/mail_message.py [hunk 2]
```diff
@@ -85,6 +85,7 @@ def default_get(self, fields):
     # content
     subject = fields.Char('Subject')
     date = fields.Datetime('Date', default=fields.Datetime.now)
+    edit_date = fields.Datetime('last edit DateTime')
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_message.py:88]: As we globally never update mesages, and as editing a message is something that is not done often, I would not define yet another field on this heavily used table, as there is already "write_date" that is available. As we have millions of message, this would require an upgrade and DB space for few added value compared to using 'write_date'. Did you find any flow that writes on the message and would lead to an invalid "write_date" ?

File: addons/mail/models/mail_thread.py [hunk 3]
```diff
@@ -4017,7 +4017,8 @@ def _message_update_content(self, message, body, attachment_ids=None,
         if strict:
             self._check_can_update_message_content(message.sudo())
 
-        msg_values = {'body': body} if body is not None else {}
+        msg_body = tools.html2plaintext(message.body)
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:4020]: Why would we do that ?

File: addons/mail/models/mail_thread.py [hunk 4]
```diff
@@ -4017,7 +4017,8 @@ def _message_update_content(self, message, body, attachment_ids=None,
...
+        msg_values = {'body': body, 'edit_date': fields.Datetime.now()} if msg_body != body else {}
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:4021]: Using the write date you would not have to change this code.

File: addons/mail/models/ir_attachment.py [hunk 5]
```diff
@@ -38,15 +38,17 @@ def register_as_main_attachment(self, force=True):
...
         for attachment in self:
             if attachment.res_model == 'mail.channel' and attachment.res_id:
                 target = self.env['mail.channel'].browse(attachment.res_id)
             else:
                 target = self.env.user.partner_id
-            self.env['bus.bus']._sendone(target, 'ir.attachment/delete', {
-                'id': attachment.id,
-            })
+            vals = {'id': attachment.id}
+            if message:
```
Comment (tde-banana-odoo) [inline addons/mail/models/ir_attachment.py:48]: Pay attention that you are doing a write in a loop, with cache mechanism this will probably lead to a single update but good practice would be to do it outside of the loop.

File: addons/mail/models/ir_attachment.py [hunk 6]
```diff
@@ -38,15 +38,17 @@ def register_as_main_attachment(self, force=True):
...
+                message.write({'edit_date': fields.Datetime.now()})
+                vals['message'] = {'id': message.id, 'edit_date': message.edit_date}
```
Comment (tde-banana-odoo) [inline addons/mail/models/ir_attachment.py:50]: You may just update vals, no rewrite it.

File: addons/mail/models/ir_attachment.py [hunk 7]
```diff
@@ -38,15 +38,17 @@ def register_as_main_attachment(self, force=True):
             with contextlib.suppress(AccessError):
                 related_record.message_main_attachment_id = self
 
-    def _delete_and_notify(self):
+    def _delete_and_notify(self, message=None):
```
Comment (tde-banana-odoo) [inline addons/mail/models/ir_attachment.py:41]: That's a bit complicated code... calling "_message_update_content" on a model with a message calls "_delete_and_notify" on attachment to do model-based code (should not be here) and to write on the message (should probably be done in the main method) ... seems to me this code could be better cut.

File: addons/mail/models/mail_thread.py (hunk 4 above)
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:4021]: Continuing previous comment: when e modify attachments, you don't update "edit_date", because it is done in a low level method ... this is a bit ping-pong code, which is hard to follow. Feel like the main method being this one, it should always be called through this one and updated in this one.

**Review by tde-banana-odoo — COMMENTED** [review]
Mail FW part, quick validation. Thanks for your work !

File: addons/mail/models/mail_message.py [hunk 8]
```diff
@@ -1085,6 +1087,21 @@ def _notify_message_notification_update(self):
         ]
         self.env['bus.bus']._sendmany(updates)
 
+    def delete_attachment_and_notify(self, attachments):
```
Comment (seb-odoo) [inline addons/mail/models/mail_message.py:1090]: method should be private (always make it so by default) as we don't want to call it externally

File: addons/mail/models/mail_thread.py (hunk 3 above)
Comment (seb-odoo) [reply addons/mail/models/mail_thread.py:4020]: Same question here, what is the reason behind this change?

File: addons/mail/models/mail_message.py [hunk 9]
```diff
@@ -1085,6 +1087,21 @@ def _notify_message_notification_update(self):
...
+        for attachment in attachments:
+            if attachment.res_model == 'mail.channel' and attachment.res_id:
+                target = self.env['mail.channel'].browse(attachment.res_id)
+            else:
+                target = self.env.user.partner_id
+            vals = {'id': attachment.id}
+            vals.update({
+                'message': {'id': self.id, 'write_date': self.write_date}
+            })
+            self.env['bus.bus']._sendone(target, 'ir.attachment/delete', vals)
```
Comment (seb-odoo) [inline addons/mail/models/mail_message.py:1102]: vals could be written as a single dict, inline of the method call

File: addons/mail/controllers/discuss.py [hunk 10]
```diff
@@ -305,19 +306,19 @@ def mail_attachment_upload(self, ufile, thread_id, thread_model, is_pending=Fals
     @http.route('/mail/attachment/delete', methods=['POST'], type='json', auth='public')
     def mail_attachment_delete(self, attachment_id, access_token=None, **kwargs):
         attachment_sudo = request.env['ir.attachment'].browse(int(attachment_id)).sudo().exists()
+        guest = request.env['mail.guest']._get_guest_from_request(request)
+        message_sudo = guest.env['mail.message'].sudo().search([('attachment_ids', 'in', attachment_sudo.ids)], limit=1)
```
Comment (seb-odoo) [inline addons/mail/controllers/discuss.py:310]: `limit=1` makes the query slower and is not necessary if you are sure there is only one message per attachment

Comment (seb-odoo) [reply addons/mail/controllers/discuss.py:310]: I see this is just moved code though, so out of scope.

File: odoo/http.py [hunk 11]
```diff
@@ -229,10 +229,6 @@ def get_default_session():
         'login': None,
         'uid': None,
         'session_token': None,
-        # profiling
-        'profile_session': None,
-        'profile_collectors': None,
-        'profile_params': None,
```
[the remark this answers is not in the document]
Comment (maki-odoo) [reply odoo/http.py]: My bad, I amend on the wrong commit

File: addons/mail/controllers/discuss.py (hunk 10 above)
Comment (seb-odoo) [inline addons/mail/controllers/discuss.py:310]: is the message guaranteed to exist/guaranteed to be found?
In chatter it's possible to have attachment linked to no message and those should be deletable as well.

File: addons/mail/static/src/core/message_model.js [hunk 12]
```diff
@@ -66,6 +67,8 @@ export class Message {
     notificationType;
     /** @type {string} */
     pinned_at;
+    /** @type {string} */
+    writeDate;
```
Comment (alexkuhn) [inline addons/mail/static/src/core/message_model.js:71]: Can just name it `write_date`, so it's the exact same naming as server-side, and you don't need to put some LOCs for renaming.

File: addons/mail/models/mail_message.py [hunk 13]
```diff
@@ -798,12 +798,11 @@ def set_message_done(self):
     @api.model
     def unstar_all(self):
         """ Unstar messages for the current partner. """
-        partner_id = self.env.user.partner_id.id
+        partner = self.env.user.partner_id
```
Comment (seb-odoo) [inline addons/mail/models/mail_message.py:801]: up to you but I feel this extra variable is not really necessary

Comment (maki-odoo) [reply addons/mail/models/mail_message.py:801]: Since I use partner in several lines, I think it is cleaner to define it as a variable.

File: addons/mail/static/src/core/message_service.js [hunk 14]
```diff
@@ -297,6 +298,11 @@ export class MessageService {
         if (data.res_model_name) {
             message.originThread.modelName = data.res_model_name;
         }
+
+        if (data.write_date) {
+            message.write_date = data.write_date;
+        }
+
```
Comment (alexkuhn) [inline addons/mail/static/src/core/message_service.js:305]: This looks already taken care by `assignDefined(message, remainingData);` above

File: addons/mail/static/src/core/message_model.js [hunk 15]
```diff
@@ -171,11 +174,12 @@ export class Message {
         return `${url("/web")}#model=${this.resModel}&id=${this.id}`;
     }
 
-    get isBodyEmpty() {
-        return (
-            !this.body ||
-            ["", "<p></p>", "<p><br></p>", "<p><br/></p>"].includes(this.body.replace(/\s/g, ""))
-        );
+    get editDate() {
+        return this.write_date !== this.date ? this.write_date : false;
```
Comment (seb-odoo) [inline addons/mail/static/src/core/message_model.js:178]: Shouldn't this compare to `create_date` rather than `date`?

All demo messages on general have edit date even though they were never edited, same for chatter messages. This might be related to this suggestion.

`date` is "unreliable" user input (especially when coming from email) whereas `create_date` and `write_date` are both technical (and thus more controlled).

File: addons/mail/static/src/core/message_model.js [hunk 16]
```diff
@@ -66,6 +67,8 @@ export class Message {
     notificationType;
     /** @type {string} */
     pinned_at;
+    /** @type {string} */
+    write_date;
```
Comment (seb-odoo) [inline addons/mail/static/src/core/message_model.js:71]: `create_date` could be documented as a field as well

**Review by alexkuhn — APPROVED** [review]
@robodoo r+

**Comment by seb-odoo** [comment]
This PR is likely the cause of https://runbot.odoo.com/runbot/build/30522486 due to no enterprise branch.

Removing merge until investigation

@robodoo r-

**Comment by alexkuhn** [comment]
@robodoo r+

File: addons/mail/models/mail_thread.py [hunk 17]
```diff
@@ -4033,7 +4033,7 @@ def _message_update_content(self, message, body, attachment_ids=None,
         if strict:
             self._check_can_update_message_content(message.sudo())
 
-        msg_values = {'body': body} if body is not None else {}
+        msg_values = {'body': body}
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:4036]: Why this change ? Anyway it does not respect the docstring (and I see no test for this change ?).

Comment (maki-odoo) [reply addons/mail/models/mail_thread.py:4036]: I am dropping this change for now. The reason of the change is, in the current webClient part, we always send the body in updating message content. It's better to change the client part in a follow-up PR.

Comment (tde-banana-odoo) [reply addons/mail/models/mail_thread.py]: Yep, better give only the values that change, like a call to ``write``, it avoids unnecessary computation / update / ...

**Comment by alexkuhn** [comment]
Need to rebase PR (some code was moved to discuss folder, from the ongoing split of code of `mail` and `discuss`)

**Comment by maki-odoo** [comment]
> Need to rebase PR (some code was moved to discuss folder, from the ongoing split of code of `mail` and `discuss`)

Done

File: addons/mail/static/src/core_ui/message.scss [hunk 18]
```diff
@@ -18,6 +18,10 @@
 
 .o-mail-Message-body > p {
     margin-bottom: 0;
+
+    &:last-of-type {
+        display: inline-block;
+    }
```
Comment (alexkuhn) [inline addons/mail/static/src/core_ui/message.scss:24]: `float` was breaking style of html content, so I kept t-outing body and just made last paragraph inline so (edited) label is inline.

**Comment by alexkuhn** [comment]
@robodoo r+
